core/trino-main/src/main/java/io/trino/sql/gen/columnar/BloomFilter.java

```diff
@@ -0,0 +1,181 @@
+        {
+            this.isNullAllowed = discreteSet.containsNull();
+            this.filter = new SliceBloomFilter((List<Slice>) (List<?>) discreteSet.getNonNullValues(), type);
+            this.inputChannels = new InputChannels(ImmutableList.of(inputChannel), ImmutableList.of(inputChannel));
+        }
+
+        @Override
+        public int filterPositionsRange(ConnectorSession session, int[] outputPositions, int offset, int size, Page page)
+        {
+            ValueBlock block = (ValueBlock) page.getBlock(0);
+            int selectedPositionsCount = 0;
+            for (int position = offset; position < offset + size; position++) {
+                boolean result = block.isNull(position) ? isNullAllowed : filter.test(block, position);
+                outputPositions[selectedPositionsCount] = position;
+                selectedPositionsCount += result ? 1 : 0;
+            }
+            return selectedPositionsCount;
+        }
+
+        @Override
+        public int filterPositionsList(ConnectorSession session, int[] outputPositions, int[] activePositions, int offset, int size, Page page)
+        {
+            ValueBlock block = (ValueBlock) page.getBlock(0);
+            int selectedPositionsCount = 0;
+            for (int index = offset; index < offset + size; index++) {
+                int position = activePositions[index];
+                boolean result = block.isNull(position) ? isNullAllowed : filter.test(block, position);
+                outputPositions[selectedPositionsCount] = position;
+                selectedPositionsCount += result ? 1 : 0;
+            }
+            return selectedPositionsCount;
+        }
+
+        @Override
+        public InputChannels getInputChannels()
+        {
+            return inputChannels;
+        }
+    }
+
+    public static final class SliceBloomFilter
+    {
+        private final long[] bloom;
+        private final int bloomSizeMask;
+        private final Type type;
+
+        /**
+         * A Bloom filter for a set of Slice values.
+         * This is approx 2X faster than the Bloom filter implementations in ORC and parquet because
+         * it uses single hash function and uses that to set 3 bits within a 64 bit word.
+         * The memory footprint is up to (4 * values.size()) bytes, which is much smaller than maintaining a hash set of strings.
+         *
+         * @param values List of values used for filtering
+         */
+        public SliceBloomFilter(List<Slice> values, Type type)
+        {
+            this.type = requireNonNull(type, "type is null");
+            int bloomSize = getBloomFilterSize(values.size());
+            bloom = new long[bloomSize];
+            bloomSizeMask = bloomSize - 1;
+            for (Slice value : values) {
+                long hashCode = XxHash64.hash(value);
+                // Set 3 bits in a 64 bit word
+                bloom[bloomIndex(hashCode)] |= bloomMask(hashCode);
```

Comment:
Did you consider using an open hash table of xxhash codes instead of the bloom filter? This could trade some performance for more accuracy.

Reply:
I want to use this eventually for collecting and evaluation dynamic filters with millions of distinct values, so I want the trade-offs to be in favor of using less memory and CPU